### ISSUE TITLE
assert.deepEqual has issues with Uint8Arrays which are formed via subarray on PhantomJS

### ISSUE THREAD

**Issue by mrstux** [chaijs/chai#943]
It is expected that assert.deepEqual() should pass a Uint8Arrays which are otherwise identical, except that one is a subarray of another. Also, this should work deeply, ie if the Uint8Arrays are properties of an object.

And this does work in most browsers, but not in PhantomJS.

The below test suite reproduces the issue, and the successful runs are copied below, as well as the PhantomJS output.

I'm seeing this issue on both Mac OS X and Windows.

The problem is that this causes lots of test suites to fail in PhantomJS for various file muxing/processing libraries, which generally function by slicing and dicing Uint8Arrays, but then in their test suites they compare them against the expected output.

As you probably know PhantomJS is used in a lot of CI systems.

Note, by 'normalizing' the sliced array by creating a new Uint8Array, you can avoid the problem, but this doesn't avoid the problem for the complex object which is composed of Uint8Arrays

A potential fix would be to normalize the a,b, when either is a Uint8Array while recursing deeply.

The following test suite can be used to reproduce/confirm the issue:

```javascript
import { assert } from 'chai';
import type from 'type-detect';


describe('Chai assert.deepEqual', function () {

	let untrimmed = new Uint8Array([
		0x01,
		0x02,
		0x03,
	]);

	it('all Uint8Arrays in tests are Uint8Array', function () {
		let trimmed = untrimmed.subarray(1, 2);
		let normalized = new Uint8Array(trimmed);

		assert.strictEqual(type(untrimmed), 'Uint8Array');
		assert.strictEqual(type(trimmed), 'Uint8Array');
		assert.strictEqual(type(normalized), 'Uint8Array');
	});

	it('trimmed, normalized and expected Uint8Arrays have same length', function () {
		let trimmed = untrimmed.subarray(1, 2);
		let normalized = new Uint8Array(trimmed);
		let expected = new Uint8Array([0x02]);

		assert.strictEqual(trimmed.length, 1);
		assert.strictEqual(normalized.length, 1);
		assert.strictEqual(expected.length, 1);
	});

	it('array contents should be the same', function () {
		let trimmed = untrimmed.subarray(1);
		let expected = new Uint8Array([0x02, 0x03]);

		assert.strictEqual(trimmed.length, expected.length );

		for( let i = 0; i < trimmed.length; i++ ) {
			assert.strictEqual( trimmed[i], expected[i]);
		}
	});

	it('should pass two identical Uint8Arrays', function () {
		let expected = new Uint8Array([
			0x01, 0x02, 0x03,
		]);

		assert.deepEqual(untrimmed, expected);
	});

	it('should pass two identical Uint8Arrays if one is a subarray of a larger array (trim beginning)', function () {
		let trimmed = untrimmed.subarray(1);
		let expected = new Uint8Array([0x02, 0x03]);

		assert.deepEqual(new Uint8Array(trimmed), expected);	// normalized works
		assert.deepEqual(trimmed, expected);					// FAIL on PhantomJS
	});

	it('should pass two identical Uint8Arrays if one is a subarray of a larger array (trim end)', function () {
		let trimmed = untrimmed.subarray(0, 2);
		let expected = new Uint8Array([0x01, 0x02]);

		assert.deepEqual(new Uint8Array(trimmed), expected);	// normalized works
		assert.deepEqual(trimmed, expected);					// FAIL on PhantomJS
	});

	it('should pass two identical Uint8Arrays if one is a subarray of a larger array (trim both ends)', function () {
		let trimmed = untrimmed.subarray(1, 2);
		let expected = new Uint8Array([0x02]);

		assert.deepEqual(new Uint8Array(trimmed), expected);	// normalized works
		assert.deepEqual(trimmed, expected);					// FAIL on PhantomJS
	});

	it('deepEqual should pass two identical Uint8Arrays if one is a subarray of a larger array (no trim)', function () {
		let trimmed = untrimmed.subarray(0, 3);
		let expected = new Uint8Array([0x01, 0x02, 0x03]);

		assert.deepEqual(trimmed, expected);	// passes, but that's expected
	});


	it('deepEqual should pass two identical prop bags made of identical Uint8Arrays', function () {
		let a = untrimmed.subarray(1);
		let b = untrimmed.subarray(0, 2);
		let c = untrimmed.subarray(1, 2);
		let d = untrimmed;

		let complexObject = { a, b, c, d };

		let expectedObject = {
			a : new Uint8Array(a),
			b : new Uint8Array(b),
			c : new Uint8Array(c),
			d : new Uint8Array(d),
		};

		assert.deepEqual( complexObject, expectedObject); // FAIL on PhantomJS
	});
});

```

**Successful tests, including tool versions,  for node, safari, firefox and chrome tests (via Node and Karma)**

```
Node Test
$ node --version
v7.7.2
$ mocha --version
3.2.0
$ mocha --compilers js:babel-register testChaiDeepEqual.js 

  Chai assert.deepEqual
    ✓ all Uint8Arrays in tests are Uint8Array
    ✓ trimmed, normalized and expected Uint8Arrays have same length
    ✓ array contents should be the same
    ✓ should pass two identical Uint8Arrays
    ✓ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim beginning)
    ✓ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim end)
    ✓ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim both ends)
    ✓ deepEqual should pass two identical Uint8Arrays if one is a subarray of a larger array (no trim)
    ✓ deepEqual should pass two identical prop bags made of identical Uint8Arrays


  9 passing (1s)


$ karma --version
Karma version: 1.3.0

$ karma start --single-run --browsers Firefox,Chrome,Safari

START:
17 03 2017 14:44:45.241:INFO [framework.browserify]: bundle built
17 03 2017 14:44:45.282:INFO [karma]: Karma v1.3.0 server started at http://localhost:9876/
17 03 2017 14:44:45.282:INFO [launcher]: Launching browsers Firefox, Chrome, Safari with unlimited concurrency
17 03 2017 14:44:45.290:INFO [launcher]: Starting browser Firefox
17 03 2017 14:44:45.314:INFO [launcher]: Starting browser Chrome
17 03 2017 14:44:45.320:INFO [launcher]: Starting browser Safari
17 03 2017 14:44:46.387:INFO [Chrome 56.0.2924 (Mac OS X 10.11.6)]: Connected on socket /#xgB0shnMGbSKEc5dAAAA with id 18352261
17 03 2017 14:44:48.432:INFO [Firefox 51.0.0 (Mac OS X 10.11.0)]: Connected on socket /#9LQlb9hs98dAnaLCAAAB with id 85074870
17 03 2017 14:45:01.168:INFO [Safari 10.0.3 (Mac OS X 10.11.6)]: Connected on socket /#o1O5OpKywWX6gjQmAAAC with id 8027771
  Chai assert.deepEqual
    ✔ all Uint8Arrays in tests are Uint8Array
    ✔ trimmed, normalized and expected Uint8Arrays have same length
    ✔ array contents should be the same
    ✔ should pass two identical Uint8Arrays
    ✔ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim beginning)
    ✔ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim end)
    ✔ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim both ends)
    ✔ deepEqual should pass two identical Uint8Arrays if one is a subarray of a larger array (no trim)
    ✔ deepEqual should pass two identical prop bags made of identical Uint8Arrays

Finished in 0.083 secs / 0.03 secs

SUMMARY:
✔ 27 tests completed
```

**yet, when run in PhantomJS**

```

$ PhantomJS --version
2.1.1

$ karma start --single-run --browsers PhantomJS

START:
17 03 2017 15:12:49.077:INFO [framework.browserify]: bundle built
17 03 2017 15:12:49.121:INFO [karma]: Karma v1.3.0 server started at http://localhost:9876/
17 03 2017 15:12:49.121:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
17 03 2017 15:12:49.126:INFO [launcher]: Starting browser PhantomJS
17 03 2017 15:12:49.701:INFO [PhantomJS 2.1.1 (Mac OS X 0.0.0)]: Connected on socket /#0-uxLU3jIudnR3zVAAAA with id 98682509
  Chai assert.deepEqual
    ✔ all Uint8Arrays in tests are Uint8Array
    ✔ trimmed, normalized and expected Uint8Arrays have same length
    ✔ array contents should be the same
    ✔ should pass two identical Uint8Arrays
    ✖ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim beginning)
    ✖ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim end)
    ✖ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim both ends)
    ✔ deepEqual should pass two identical Uint8Arrays if one is a subarray of a larger array (no trim)
    ✖ deepEqual should pass two identical prop bags made of identical Uint8Arrays

Finished in 0.024 secs / 0.013 secs

SUMMARY:
✔ 5 tests completed
✖ 4 tests failed

FAILED TESTS:
  Chai assert.deepEqual
    ✖ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim beginning)
      PhantomJS 2.1.1 (Mac OS X 0.0.0)
    expected { '0': 2,
      '1': 3,
      length: 2,
      byteOffset: 1,
      buffer: { byteLength: 3, slice: [Function: slice] },
      byteLength: 2,
      subarray: [Function: subarray],
      BYTES_PER_ELEMENT: 1,
      set: [Function: set] } to deeply equal { '0': 2,
      '1': 3,
      length: 2,
      byteOffset: 0,
      buffer: { byteLength: 2, slice: [Function: slice] },
      byteLength: 2,
      subarray: [Function: subarray],
      BYTES_PER_ELEMENT: 1,
      set: [Function: set] }
    AssertionError@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:592:24
    assert@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:2696:31
    assertEql@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:3299:16
    /var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:6682:30
    deepEqual@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:4857:35
    /var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:136609:25

    ✖ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim end)
      PhantomJS 2.1.1 (Mac OS X 0.0.0)
    expected { '0': 1,
      '1': 2,
      length: 2,
      byteOffset: 0,
      buffer: { byteLength: 3, slice: [Function: slice] },
      byteLength: 2,
      subarray: [Function: subarray],
      BYTES_PER_ELEMENT: 1,
      set: [Function: set] } to deeply equal { '0': 1,
      '1': 2,
      length: 2,
      byteOffset: 0,
      buffer: { byteLength: 2, slice: [Function: slice] },
      byteLength: 2,
      subarray: [Function: subarray],
      BYTES_PER_ELEMENT: 1,
      set: [Function: set] }
    AssertionError@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:592:24
    assert@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:2696:31
    assertEql@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:3299:16
    /var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:6682:30
    deepEqual@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:4857:35
    /var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:136617:25

    ✖ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim both ends)
      PhantomJS 2.1.1 (Mac OS X 0.0.0)
    expected { '0': 2,
      length: 1,
      byteOffset: 1,
      buffer: { byteLength: 3, slice: [Function: slice] },
      byteLength: 1,
      subarray: [Function: subarray],
      BYTES_PER_ELEMENT: 1,
      set: [Function: set] } to deeply equal { '0': 2,
      length: 1,
      byteOffset: 0,
      buffer: { byteLength: 1, slice: [Function: slice] },
      byteLength: 1,
      subarray: [Function: subarray],
      BYTES_PER_ELEMENT: 1,
      set: [Function: set] }
    AssertionError@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:592:24
    assert@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:2696:31
    assertEql@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:3299:16
    /var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:6682:30
    deepEqual@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:4857:35
    /var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:136624:25

    ✖ deepEqual should pass two identical prop bags made of identical Uint8Arrays
      PhantomJS 2.1.1 (Mac OS X 0.0.0)
    expected { Object (a, b, ...) } to deeply equal { Object (a, b, ...) }
    AssertionError@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:592:24
    assert@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:2696:31
    assertEql@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:3299:16
    /var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:6682:30
    deepEqual@/var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:4857:35
    /var/folders/1r/2d585xg13gb22j0jp6t3fktc0000gn/T/3aefe315dbea3bab93accbc22f328f92.browserify:136649:25

```

The problem is that this causes lots of test suites to fail in PhantomJS for various file muxing/processing libraries, which generally function by slicing and dicing Uint8Arrays, but then in their test suites they compare them against the expected output.

Note, by 'normalizing' the sliced array by creating a new Uint8Array, you can avoid the problem, but this doesn't avoid the problem for the complex object which is composed of Uint8Arrays

**Comment by meeber** [comment]
@mrstux Thanks for the issue. Few questions:

1. Does this problem also happen in Chai v4? (Not officially released but available on npm via `chai@canary`)
2. Are you suggesting that PhantomJS is doing the correct thing and that Chai is doing something wrong, so Chai needs to be fixed? Or are you suggesting that PhantomJS is doing the wrong thing and that Chai is technically doing the right thing, but that Chai should be updated with special logic to compensate for PhantomJS doing the wrong thing since PhantomJS is commonly used?

**Comment by mrstux** [comment]
@meeber 

1. it works as expected in `chai@canary`

```
npm list chai
<redacted>@1.14.0 /Users/stux/git/<redacted>
└── chai@4.0.0-canary.1 

stux$ karma start karma-local.conf.js --browsers PhantomJS
23 03 2017 13:31:48.992:INFO [framework.browserify]: registering rebuild (autoWatch=true)
23 03 2017 13:31:49.000:INFO [framework.detect-browsers]: Detecting browsers is disabled. The browsers of the browsers array are used.
23 03 2017 13:32:04.355:INFO [framework.browserify]: 7106918 bytes written (11.47 seconds)
23 03 2017 13:32:04.383:INFO [framework.browserify]: bundle built
23 03 2017 13:32:04.423:WARN [karma]: No captured browser, open http://localhost:9876/
23 03 2017 13:32:04.431:INFO [karma]: Karma v1.3.0 server started at http://localhost:9876/
23 03 2017 13:32:04.431:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
23 03 2017 13:32:04.439:INFO [launcher]: Starting browser PhantomJS
23 03 2017 13:32:04.989:INFO [PhantomJS 2.1.1 (Mac OS X 0.0.0)]: Connected on socket /#_DP7qm6VOTzmZtOTAAAA with id 76827952

START:
  Chai assert.deepEqual
    ✔ all Uint8Arrays in tests are Uint8Array
    ✔ trimmed, normalized and expected Uint8Arrays have same length
    ✔ array contents should be the same
    ✔ should pass two identical Uint8Arrays
    ✔ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim beginning)
    ✔ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim end)
    ✔ should pass two identical Uint8Arrays if one is a subarray of a larger array (trim both ends)
    ✔ deepEqual should pass two identical Uint8Arrays if one is a subarray of a larger array (no trim)
    ✔ deepEqual should pass two identical prop bags made of identical Uint8Arrays

Finished in 0.018 secs / 0.011 secs

SUMMARY:
✔ 9 tests completed
```

2. I'm not sure who is doing the wrong thing. Either way, semantically, the UInt8Arrays are the same, the difference appears to be in the backing array buffer, which I believe is prototypical. Since chai@canary does work, that strongly hints that perhaps chain@^3.5.0 should also pass this test, assuming its still being maintained

The version of chai which we use is ^3.5.0

```
└─┬ chai@3.5.0 
  ├─┬ deep-eql@0.1.3 
  │ └── type-detect@0.1.1 
  └── type-detect@1.0.0 
```

**Comment by mrstux** [comment]
The version of deep-eql that we have adopted is the latest version, 2.0.1, and this version resolves other uint8array comparisons that were failing in assert.deepEqual

├─┬ chai@3.5.0
│ └── deep-eql@0.1.3 
└── deep-eql@2.0.1

**Comment by mrstux** [comment]
Similar issues occurred with type-detect as well

├─┬ chai@3.5.0
│ ├─┬ deep-eql@0.1.3
│ │ └── type-detect@0.1.1 
│ └── type-detect@1.0.0 
├─┬ deep-eql@2.0.1
│ └── type-detect@3.0.0 
└── type-detect@4.0.0

**Comment by meeber** [comment]
Unfortunately we don't currently have the cycles to backport fixes in master to v3.5. Closing this since it's resolved in master. Thanks!
